source/arm11/patch.c

```diff
@@ -0,0 +1,367 @@
+#include <stdlib.h>
+#include <string.h>
+#include "types.h"
+#include "util.h"
+#include "arm11/drivers/hid.h"
+#include "drivers/lgy.h"
+#include "arm11/fmt.h"
+#include "fs.h"
+#include "arm11/drivers/lcd.h"
+#include "arm11/patch.h"
+#include "arm11/power.h"
+
+const Result RES_PATCH_INVALID = 202u; //should move to error_codes.h in libn3ds
+
+typedef struct {
+    u8 *buffer;
+    u16 cacheSize;
+    u16 cacheOffset;
+    u16 maxCacheSize;
+} Cache;
+
+static u8 readCache(const FHandle patchHandle, Cache *cache, Result *res) {
+    u8 result = (cache->buffer)[(cache->cacheOffset)++];
+    if((cache->cacheOffset) >= (cache->cacheSize)) {
+        (cache->cacheSize) = min((cache->maxCacheSize), ((fSize(patchHandle)-12) - fTell(patchHandle)));
+        *res = fRead(patchHandle, (cache->buffer), (cache->cacheSize), NULL);
+        (cache->cacheOffset) = 0;
+    }
+
+    return result;
+}
+
+static Result patchIPS(const FHandle patchHandle) {
+	Result res = RES_OK;
+
+	ee_puts("IPS patch found! Patching...");
+
+	//verify patch is IPS (magic number is "PATCH")
+	bool isValidPatch = false;
+	char *magicBuf = (char*)calloc(5, 1);
```

Comment:
Using calloc() for very small buffers is not a good idea in my opinion. It's much slower than stack buffers and fragments the heap.